cdc_iterator.go

```diff
@@ -0,0 +1,293 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-mysql/common"
+	"github.com/conduitio/conduit-commons/csync"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-mysql-org/go-mysql/canal"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/schema"
+	"gopkg.in/tomb.v2"
+)
+
+type cdcIterator struct {
+	canal *canal.Canal
+	acks  *csync.WaitGroup
+	data  chan *canal.RowsEvent
+
+	// canalTomb is a tomb that simplifies handling canal run errors
+	canalTomb *tomb.Tomb
+
+	config cdcIteratorConfig
+}
+
+type cdcIteratorConfig struct {
+	common.SourceConfig
+	position  sdk.Position
+	TableKeys common.TableKeys
+}
+
+func newCdcIterator(ctx context.Context, config cdcIteratorConfig) (common.Iterator, error) {
+	c, err := common.NewCanal(config.SourceConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	iterator := &cdcIterator{
+		canal:     c,
+		acks:      &csync.WaitGroup{},
+		data:      make(chan *canal.RowsEvent),
+		canalTomb: &tomb.Tomb{},
+		config:    config,
+	}
+
+	startPosition, err := iterator.getStartPosition(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start position: %w", err)
+	}
+
+	iterator.canalTomb.Go(func() error {
```

Comment:
We use tomb to manage a single goroutine, it feels like this could be simplified.

Reply:
Yup. Pushed new changes replacing its usage by 2 channels, lmk what you think about it.

Reply:
Yeah, definitely simpler 👍